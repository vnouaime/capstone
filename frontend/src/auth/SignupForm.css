.signupform-header {
    margin-top: 30px;
}

.signupform-card {
    border-color: rgb(137, 225, 169);
}

.signupform-input {
    border-color: rgb(137, 225, 169);
}

.signupform-button {
    border-color:rgb(137, 225, 169);
    background-color: rgb(137, 225, 169);
    color: white;
}