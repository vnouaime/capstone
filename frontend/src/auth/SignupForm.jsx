import React, { useState } from "react";
import { useNavigate, Link } from "react-router-dom";
import { Form, Button, Container, Card, Col } from "react-bootstrap";

import Alert from "../common/Alert";

import "./SignupForm.css"

/** Signup form.
 *
 * Renders form for signing up a new user. 
 * Changes state as user types in form. 
 * When submitted, calls signup function passed from prop. 
 * If successful, redirects to /. If not, 
 * alerts user with message. 
 * 
 * Props: 
 * signup
 * 
 * State: 
 * formData, formErrors
 *
 * Routes -> SignupForm -> Alert
 * Routed as /signup
 */
const SignupForm = ({ signup }) => {
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: "",
  });
  const [formErrors, setFormErrors] = useState([]);

  async function handleSubmit(evt) {
    evt.preventDefault();
    let result = await signup(formData);

    if (result.success) {
      navigate("/");
    } else {
        setFormErrors(result.errors);
      }
  }

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData(data => ({ ...data, [name]: value }));
  }

  return (
    <Container className="SignupForm">
      <Container className="col-md-10 offset-md-1 col-lg-8 offset-lg-2 mb-5">
        <h1 className="signupform-header mb-3 text-center">Sign Up</h1>
        {formErrors.length ? (
                  <Alert type="danger" messages={formErrors} />
                ) : null}
          <Card className="signupform-card">
            <Card.Body>
              <Form onSubmit={handleSubmit}>
                <Form.Group className="mb-3">
                  <Form.Label htmlFor="username">Username</Form.Label>
                  <Form.Control
                    className="signupform-input"
                    id="username"
                    name="username"
                    type="text"
                    placeholder="Username"
                    value={formData.username}
                    onChange={handleChange}
                    required
                  />
                </Form.Group>
                <Form.Group className="mb-3">
                  <Form.Label htmlFor="password">Password</Form.Label>
                  <Form.Control
                    className="signupform-input"
                    id="password"
                    name="password"
                    type="password"
                    placeholder="Password"
                    value={formData.password}
                    onChange={handleChange}
                    required
                    minLength="5"
                  />
                </Form.Group>
                <Form.Group className="mb-3">
                  <Form.Label htmlFor="firstName">First Name</Form.Label>
                  <Form.Control
                    className="signupform-input"
                    id="firstName"
                    name="firstName"
                    type="text"
                    placeholder="First Name"
                    value={formData.firstName}
                    onChange={handleChange}
                    required
                  />
                </Form.Group>
                <Form.Group className="mb-3">
                  <Form.Label htmlFor="lastName">Last Name</Form.Label>
                  <Form.Control
                    className="signupform-input"
                    id="lastName"
                    name="lastName"
                    type="text"
                    placeholder="Last Name"
                    value={formData.lastName}
                    onChange={handleChange}
                    required
                  />
                </Form.Group>
                <Form.Group className="mb-3" as={Col} controlId="formGridEmail">
                  <Form.Label htmlFor="email">Email</Form.Label>
                  <Form.Control
                    className="signupform-input"
                    id="email"
                    name="email"
                    type="email"
                    placeholder="Email"
                    value={formData.email}
                    onChange={handleChange}
                    required
                  />
                </Form.Group>
                <div className="d-flex justify-content-end">
                  <Button className="signupform-button" type="submit" data-testid="signup-button">
                    Submit
                  </Button>
                </div>
              </Form>
            </Card.Body>
          </Card>
      </Container>
      <div className="text-center">Already have an account? <Link to="/login">Log in here</Link></div>
    </Container>
  );
}

export default SignupForm;
