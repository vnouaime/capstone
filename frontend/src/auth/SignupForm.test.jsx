import React from "react";
import { expect, vi } from "vitest"
import { render, fireEvent } from "vitest-react";
import { MemoryRouter } from "react-router-dom";
import SignupForm from "./SignupForm";

test("smoke test", () => {
    render(
        <MemoryRouter>
            <SignupForm />
        </MemoryRouter>
    );
});

test("snapshot", () => {
    const { asFragment } = render(<MemoryRouter><SignupForm /></MemoryRouter>)
    expect(asFragment()).toMatchSnapshot()
})

test("should signup user", async () => {
    const signup = vi.fn().mockResolvedValue({ success: true })

    const { getByTestId, getByLabelText } = 
        render(
            <MemoryRouter>
                <SignupForm signup={signup}/>
            </MemoryRouter>)

    const username = getByLabelText("Username")
    const password = getByLabelText("Password")
    const firstName = getByLabelText("First Name")
    const lastName = getByLabelText("Last Name")
    const email = getByLabelText("Email")
    const signupButton = getByTestId("signup-button")
    
    fireEvent.change(username, { target: { value: 'testuser' } })
    fireEvent.change(password, { target: { value: 'password' } })
    fireEvent.change(firstName, { target: { value: 'Test' } })
    fireEvent.change(lastName, { target: { value: 'Test' } })
    fireEvent.change(email, { target: { value: 'test@test.com' } })
    fireEvent.click(signupButton)

    await vi.waitFor(() => expect(window.location.pathname).toBe('/'));
})

test("invalid signup: duplicate username", async () => {
    const signup = vi.fn().mockResolvedValue({ success: false, errors: ["Username taken: 'testuser'"]})

    const { getByTestId, getByLabelText, findByText } = 
        render(
            <MemoryRouter>
                <SignupForm signup={signup}/>
            </MemoryRouter>)

    const username = getByLabelText("Username")
    const password = getByLabelText("Password")
    const firstName = getByLabelText("First Name")
    const lastName = getByLabelText("Last Name")
    const email = getByLabelText("Email")
    const signupButton = getByTestId("signup-button")

    fireEvent.change(username, { target: { value: 'testuser' } })
    fireEvent.change(password, { target: { value: 'password' } })
    fireEvent.change(firstName, { target: { value: 'Test' } })
    fireEvent.change(lastName, { target: { value: 'Test' } })
    fireEvent.change(email, { target: { value: 'test@test.com' } })
    fireEvent.click(signupButton)

    const alert = await findByText("Username taken: 'testuser'");
    expect(alert).toBeInTheDocument();
    expect(alert.closest('div')).toHaveClass("alert-danger");
})


