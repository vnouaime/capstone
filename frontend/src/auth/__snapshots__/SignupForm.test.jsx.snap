// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot 1`] = `
<DocumentFragment>
  <div
    class="SignupForm container"
  >
    <div
      class="col-md-10 offset-md-1 col-lg-8 offset-lg-2 mb-5 container"
    >
      <h1
        class="signupform-header mb-3 text-center"
      >
        Sign Up
      </h1>
      <div
        class="signupform-card card"
      >
        <div
          class="card-body"
        >
          <form
            class=""
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="username"
              >
                Username
              </label>
              <input
                class="signupform-input form-control"
                id="username"
                name="username"
                placeholder="Username"
                required=""
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="password"
              >
                Password
              </label>
              <input
                class="signupform-input form-control"
                id="password"
                minlength="5"
                name="password"
                placeholder="Password"
                required=""
                type="password"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="firstName"
              >
                First Name
              </label>
              <input
                class="signupform-input form-control"
                id="firstName"
                name="firstName"
                placeholder="First Name"
                required=""
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="lastName"
              >
                Last Name
              </label>
              <input
                class="signupform-input form-control"
                id="lastName"
                name="lastName"
                placeholder="Last Name"
                required=""
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3 col"
            >
              <label
                class="form-label"
                for="email"
              >
                Email
              </label>
              <input
                class="signupform-input form-control"
                id="email"
                name="email"
                placeholder="Email"
                required=""
                type="email"
                value=""
              />
            </div>
            <div
              class="d-flex justify-content-end"
            >
              <button
                class="signupform-button btn btn-primary"
                data-testid="signup-button"
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      class="text-center"
    >
      Already have an account? 
      <a
        href="/login"
      >
        Log in here
      </a>
    </div>
  </div>
</DocumentFragment>
`;
